Parser/Types/TypeAnalyze.cs

```diff
@@ -23,16 +23,43 @@ public override NType Visit (NProgram p)
    }
 
    public override NType Visit (NDeclarations d) {
-      Visit (d.Vars); return Visit (d.Funcs);
+      Visit (d.Consts); Visit (d.Vars); return Visit (d.Funcs);
+   }
+   public override NType Visit (NConstDecl c) {
+      if (mSymbols.Consts.Any (a => a.Name.Text == c.Name.Text))
+         throw new ParseException (c.Name, "Constant with the same name already declared");
+      c.Value.Accept (this);
+      mSymbols.Consts.Add (c);
+      return c.Value.Type;
    }
-
    public override NType Visit (NVarDecl d) {
+      if (mSymbols.Consts.Any (a => a.Name.Text == d.Name.Text))
+         throw new ParseException (d.Name, "Constant with the same name already declared");
+      if (mSymbols.Vars.Any (a => a.Name.Text == d.Name.Text))
+         throw new ParseException (d.Name, "Variable with the same name already declared");
       mSymbols.Vars.Add (d);
       return d.Type;
    }
-
    public override NType Visit (NFnDecl f) {
+      if (mSymbols.Consts.Any (a => a.Name.Text == f.Name.Text))
+         throw new ParseException (f.Name, "Constant with the same function name already declared");
+      if (mSymbols.Vars.Any (a => a.Name.Text == f.Name.Text))
+         throw new ParseException (f.Name, "Variable with the same function name already declared");
+      if (mSymbols.Funcs.Any (a => a.Name.Text == f.Name.Text))
+         throw new ParseException (f.Name, "Function with the same function name already declared");

```

Comment:
Just put this code in one helper function and call it for all the 3 Visit routines above. When parsing a const decl, the Vars and Funcs lists are empty so no issues. When parsing a VarDecl, the Funcs list is empty etc...